### PR TITLE
graphql-cop now runs against a list of endpoints and continues iterating through the list until all endpoints are tested in case there is more than one if the user had not specified the path (e.g. -t http://example.com). If the user had already specified the path (e.g. -t http://example.com/graphql) then only uses this endpoint.

### PR DESCRIPTION
**Pull request by halfluke**
└─# python3 graphql-cop.py -x http://127.0.0.1:8080   -t http://localhost:5013/graphql  
[LOW] Field Suggestions - Field Suggestions are Enabled (Information Leakage)
[HIGH] Introspection - Introspection Query Enabled (Information Leakage)
[LOW] GraphQL IDE - GraphiQL Explorer/Playground Enabled (Information Leakage)
[MEDIUM] GET Method Query Support - GraphQL queries allowed using the GET method (Possible Cross Site Request Forgery (CSRF))
[HIGH] Alias Overloading - Alias Overloading with 100+ aliases is allowed (Denial of Service)
[HIGH] Array-based Query Batching - Batch queries allowed with 10+ simultaneous queries (Denial of Service)
[HIGH] Field Duplication - Queries are allowed with 500 of the same repeated field (Denial of Service)
[HIGH] Directive Overloading - Multiple duplicated directives allowed in a query (Denial of Service)
[HIGH] Introspection-based Circular Query - Circular-query using Introspection (Denial of Service)
[MEDIUM] POST based url-encoded query (possible CSRF) - GraphQL accepts non-JSON queries over POST (Possible Cross Site Request Forgery)
**[INFO] Unhandled Errors Detection - Exception errors are not handled (Information Leakage)**
                                                                                             
image

### CODE REVIEW

File: graphql-cop.py [hunk 1]
```diff
@@ -81,8 +81,14 @@
 for test in tests:
     json_output.append(test(url, proxy, HEADERS))
 
+if hasattr(detect_graphiql, 'GraphQLIDEpath'):
```
Comment (dolevf) [inline graphql-cop.py:84]: Why is this required? does the output not get appended in the for loop for these tests?

File: version.py [hunk 2]
```diff
@@ -1,2 +1,2 @@
 """Version details of graphql-cop."""
-VERSION = '1.9'
+VERSION = '1.99'
```
Comment (dolevf) [inline version.py:2]: oops

File: graphql-cop.py (hunk 1 above)
Comment (halfluke) [reply graphql-cop.py:84]: Hi, this is just to say that if a graphql ide test was found, we should also repeat the tracing and unhandled errors tests for the graphql ide path (dvga case). Then I added the url in the tool output because we could in theory have two positive results for both the "normal" url and the graphql ide url
New output:
[INFO] Unhandled Errors Detection - Exception errors are not handled (Information Leakage - http://localhost:5013/graphiql)

Comment (dolevf) [reply graphql-cop.py:84]: I suggest the following changes:

1. graphql-cop runs against a list of endpoints, continue iterating through the list until all endpoints are tested in case there is more than one. If the user already specifies the path (e.g. `-t http://example.com/graphql`) then only use this endpoint.
2. Run all test cases against the one (or more) endpoints that were found
3. Move out the list of endpoints e.g. `endpoints = ['graphiql', 'playground', 'console', 'graphql']` to a separate file that's sourced by all the relevant tests so they aren't hard coded per test
4. Bump version to `1.9.1`

Comment (halfluke) [reply graphql-cop.py:84]: Not sure... my current changes are minor:
all tests are run against the endpoint specified with -t like before, with the only addition that if the tool finds a graphql ide, it will save that endpoint and will also run stack trace/exception test against it (in addition to the endpoint specified with -t).
Are you suggesting to use the list endpoints = ['graphiql', 'playground', 'console', 'graphql'] not only for the graphql ide test, but for all tests, unless -t is specified? In your suggestion, if we use -t we would no longer be able to find a possible graphql ide because we would no longer test the whole list of possible ide endpoints...

Comment (dolevf) [reply graphql-cop.py:84]: The `-t` option is there if the user already knows that GraphQL lives at the provided path, so I'm less concerned about potentially misses. 

Use version `1.10`

Comment (halfluke) [reply graphql-cop.py:84]: so how would you like the user to provide the url? Just like http://urlhere:port ? With a different switch, such as -u ?
I had this test where graphql was at https://urlhere/cz/graphq. In this case, would it be -u https://urlhere/cz and then the tool would iterate through all the list of endpoints?

Comment (dolevf) [reply graphql-cop.py:84]: The same way it's done today:
```
python3 graphql-cop.py  -t http://localhost:5013/graphql
```
if a path is provided - you check if GraphQL is present at `/graphql`/

If its ran this way:
```
python3 graphql-cop.py  -t http://localhost:5013
```

Such as that the path is not provided at all - you first check the webroot (`/`) and then iterate through the various possible endpoints until the API is found

Comment (halfluke) [reply graphql-cop.py:84]: image

if you like this output I will commit.
What version should I update to?

Comment (dolevf) [reply graphql-cop.py:84]: looks ok, but please remove the URL from the parentheses.

`1.10`

Comment (halfluke) [reply graphql-cop.py:84]: but the url is the main point because it specifies the exact path for the issue!
As you can see in the first example, there are some duplicates that are different only for the url, because the tool iterates across different endpoints and may find an issue only on a single endpoint, and another issue on more than one.

Comment (dolevf) [reply graphql-cop.py:84]: Use only the path instead of the full string

**Comment by halfluke** [comment]
Please let me know if everything is ok for you now. Thanks

**Review by dolevf — APPROVED** [review]
Gave it a try - looks good!

**Comment by dolevf** [comment]
Merged, thank you for making this tool a lot better.

**Comment by halfluke** [comment]
My pleasure!
